### PR TITLE
[client,sdl] encapsulate c++ context

### PR DESCRIPTION
**Pull request by akallabeth**
to properly initialize c++ us a heap allocated wrapper class

### CODE REVIEW

**Review by hardening — APPROVED** [review]
Looks ok

File: client/SDL/sdl_types.hpp [hunk 1]
```diff
@@ -19,4 +19,28 @@
...
+{
+	rdpClientContext common;
+	SdlContext* sdl;
+} sdl_rdp_context;
+
+static inline SdlContext* get_context(void* ctx)
+{
+	if (!ctx)
+		return nullptr;
+	auto sdl = reinterpret_cast<sdl_rdp_context*>(ctx);
```
Comment (mfleisz) [inline client/SDL/sdl_types.hpp:36]: Should be `static_cast`

File: client/SDL/sdl_monitor.cpp [hunk 2]
```diff
@@ -99,13 +99,13 @@ static BOOL sdl_is_monitor_id_active(sdlContext* sdl, UINT32 id)
...
 
-static BOOL sdl_apply_max_size(sdlContext* sdl, UINT32* pMaxWidth, UINT32* pMaxHeight)
+static BOOL sdl_apply_max_size(SdlContext* sdl, UINT32* pMaxWidth, UINT32* pMaxHeight)
 {
 	WINPR_ASSERT(sdl);
 	WINPR_ASSERT(pMaxWidth);
 	WINPR_ASSERT(pMaxHeight);
 
-	rdpSettings* settings = sdl->common.context.settings;
+	rdpSettings* settings = sdl->context()->settings;
```
Comment (mfleisz) [inline client/SDL/sdl_monitor.cpp:108]: Use `auto` (same in following functions)

File: client/SDL/sdl_freerdp.cpp [hunk 3]
```diff
@@ -301,33 +301,33 @@ static BOOL sdl_begin_paint(rdpContext* context)
 	return TRUE;
 }
 
-static BOOL sdl_redraw(sdlContext* sdl)
+static BOOL sdl_redraw(SdlContext* sdl)
 {
 	WINPR_ASSERT(sdl);
 
-	rdpGdi* gdi = sdl->common.context.gdi;
+	rdpGdi* gdi = sdl->context()->gdi;
```
Comment (mfleisz) [inline client/SDL/sdl_freerdp.cpp:308]: Use `auto`

File: client/SDL/sdl_freerdp.cpp [hunk 4]
```diff
@@ -383,7 +382,7 @@ static BOOL sdl_end_paint_process(rdpContext* context)
 		}
 		else
 		{
-			const Uint32 id = SDL_GetWindowID(window->window);
+			const Uint32 id = SDL_GetWindowID(window.window);
```
Comment (mfleisz) [inline client/SDL/sdl_freerdp.cpp:385]: `auto`

File: client/SDL/sdl_freerdp.cpp [hunk 5]
```diff
@@ -592,44 +590,36 @@ static const char* sdl_window_get_title(rdpSettings* settings)
...
 	SDL_Quit();
 }
 
-static BOOL sdl_create_windows(sdlContext* sdl)
+static BOOL sdl_create_windows(SdlContext* sdl)
 {
 	WINPR_ASSERT(sdl);
 
-	const char* title = sdl_window_get_title(sdl->common.context.settings);
+	const char* title = sdl_window_get_title(sdl->context()->settings);
```
Comment (mfleisz) [inline client/SDL/sdl_freerdp.cpp:611]: `auto`

File: client/SDL/sdl_freerdp.cpp [hunk 6]
```diff
@@ -639,32 +629,31 @@ static BOOL sdl_create_windows(sdlContext* sdl)
...
+	sdl->windows_created.set();
 	return rc;
 }
 
-static BOOL sdl_wait_create_windows(sdlContext* sdl)
+static BOOL sdl_wait_create_windows(SdlContext* sdl)
 {
-	std::lock_guard<CriticalSection> lock(*sdl->critical);
-	sdl->windows_created->clear();
+	std::lock_guard<CriticalSection> lock(sdl->critical);
```
Comment (mfleisz) [inline client/SDL/sdl_freerdp.cpp:651]: Since we use C++17 just use `std::lock_guard lock(...);`

File: client/SDL/sdl_freerdp.cpp [hunk 7]
```diff
@@ -692,30 +681,28 @@ static int sdl_run(sdlContext* sdl)
...
-		while (!freerdp_shall_disconnect_context(&sdl->common.context) &&
-		       SDL_WaitEvent(&windowEvent))
+		while (!freerdp_shall_disconnect_context(sdl->context()) && SDL_WaitEvent(&windowEvent))
 		{
 #if defined(WITH_DEBUG_SDL_EVENTS)
 			SDL_Log("got event %s [0x%08" PRIx32 "]", sdl_event_type_str(windowEvent.type),
 			        windowEvent.type);
 #endif
-			std::lock_guard<CriticalSection> lock(*sdl->critical);
+			std::lock_guard<CriticalSection> lock(sdl->critical);
```
Comment (mfleisz) [inline client/SDL/sdl_freerdp.cpp:695]: `std::lock_guard` is enough

File: client/SDL/sdl_freerdp.cpp [hunk 8]
```diff
@@ -959,31 +935,26 @@ static void sdl_post_final_disconnect(freerdp* instance)
...
-	auto sdl = static_cast<sdlContext*>(arg);
 	DWORD nCount;
 	DWORD status;
 	int exit_code = SDL_EXIT_SUCCESS;
 	HANDLE handles[MAXIMUM_WAIT_OBJECTS] = {};
 
 	WINPR_ASSERT(sdl);
 
-	freerdp* instance = sdl->common.context.instance;
+	freerdp* instance = sdl->context()->instance;
```
Comment (mfleisz) [inline client/SDL/sdl_freerdp.cpp:952]: `auto`

File: client/SDL/sdl_freerdp.cpp [hunk 9]
```diff
@@ -1290,35 +1253,49 @@ int main(int argc, char* argv[])
 	return rc;
 }
 
-BOOL sdl_context::update_fullscreen(BOOL enter)
+BOOL SdlContext::update_fullscreen(BOOL enter)
 {
-	std::lock_guard<CriticalSection> lock(*critical);
-	for (uint32_t x = 0; x < windowCount; x++)
+	std::lock_guard<CriticalSection> lock(critical);
```
Comment (mfleisz) [inline client/SDL/sdl_freerdp.cpp:1258]: `std::lock_guard lock(...);`

File: client/SDL/sdl_freerdp.cpp [hunk 10]
```diff
@@ -1290,35 +1253,49 @@ int main(int argc, char* argv[])
...
 	}
 	fullscreen = enter;
 	return TRUE;
 }
 
-BOOL sdl_context::update_resizeable(BOOL enable)
+BOOL SdlContext::update_resizeable(BOOL enable)
 {
-	std::lock_guard<CriticalSection> lock(*critical);
+	std::lock_guard<CriticalSection> lock(critical);
```
Comment (mfleisz) [inline client/SDL/sdl_freerdp.cpp:1270]: `std::lock_guard lock(...);`

File: client/SDL/sdl_freerdp.cpp [hunk 11]
```diff
@@ -1290,35 +1253,49 @@ int main(int argc, char* argv[])
 	return rc;
 }
 
-BOOL sdl_context::update_fullscreen(BOOL enter)
+BOOL SdlContext::update_fullscreen(BOOL enter)
 {
-	std::lock_guard<CriticalSection> lock(*critical);
-	for (uint32_t x = 0; x < windowCount; x++)
+	std::lock_guard<CriticalSection> lock(critical);
+	for (auto& window : windows)
```
Comment (mfleisz) [inline client/SDL/sdl_freerdp.cpp:1259]: Should be `const auto&`

File: client/SDL/sdl_freerdp.cpp [hunk 12]
```diff
@@ -1290,35 +1253,49 @@ int main(int argc, char* argv[])
...
+	std::lock_guard<CriticalSection> lock(critical);
 
-	const rdpSettings* settings = common.context.settings;
+	const auto settings = context()->settings;
 	const BOOL dyn = freerdp_settings_get_bool(settings, FreeRDP_DynamicResolutionUpdate);
 	const BOOL smart = freerdp_settings_get_bool(settings, FreeRDP_SmartSizing);
 	BOOL use = (dyn && enable) || smart;
 
-	for (uint32_t x = 0; x < windowCount; x++)
+	for (auto& window : windows)
```
Comment (mfleisz) [inline client/SDL/sdl_freerdp.cpp:1277]: Should be `const auto&`

**Review by mfleisz — COMMENTED** [review]
A few minor comments - overall LGTM

File: client/SDL/sdl_freerdp.cpp (hunk 7 above)
Comment (akallabeth) [reply client/SDL/sdl_freerdp.cpp:695]: @mfleisz sadly our winodws builder is not up to date enough for that syntax.

Comment (mfleisz) [reply client/SDL/sdl_freerdp.cpp:695]: @akallabeth ah yeah - somehow I had in mind we require C++17 (which VC2003 certainly is not capable of) 😅

**Review by mfleisz — APPROVED** [review]

**Pull request merged**